IBX-4996: Storefront tutorial (#1910)
ezsystems/developer-documentation: docs/commerce/cart/cart.md

There is no specific configuration related to the cart component.
All configuration is done at the checkout and storefront level.

> Cart constructor takes the following argument:

Reviewer: ```suggestion
Cart constructor takes the following arguments:
```

Main cart is the main user interface component of the cart.
By default, it lists items selected for purchase and requested cart item quantities. 

> Users who view the Main cart can remove individual items.

Reviewer 2: I think we want to hint that we are talking about a specific component.

> Users who view the Main cart can remove individual items.

Author: Yes, I agree with Kuba. In this context it should be capitalized.

You could extend the Cart summary widget to let buyers navigate from this view, 
for example, to checkout, or back to shopping, by adding respective buttons.

### Cart events

When cart data is changed or loaded, the `ibexa-cart:cart-data-changed` event is triggered on `body`.

> The reference to the Cart is sent in the event's `detail`.

Reviewer: I like that as a general rule: backticks and capital when speaking about specific object, lowercase and no backticks when talking in general about functionality, what do you say, @juskora ?

> The reference to the Cart is sent in the event's `detail`.

Author: I think when talking about the specific object, we should capitalized the word, without putting inside backticks.

```js
document.body.addEventListener(
    'ibexa-cart:cart-data-changed',
    ({ detail: { cart } }) => {
        refreshMyWidget(cart);
    },
    false,
);
```

### Cart service 

The Cart package provides `Ibexa\Contracts\Cart\CartServiceInterface` Symfony service, 
which is the entrypoint for calling the [backend API](cart_api.md).

You can import services using the following code:

```js
import * as cartService from '@ibexa-cart/src/bundle/Resources/public/js/service/cart';
```

Next, use in the code as follows

```js
cartService.deleteCartEntry(cartIdentifier, entryIdentifier);
```

> Every cart service function returns a Promise with a parsed response.

Reviewer 2: Good point, they actually use lowercase _promise_ also. I like their style of using both. They mention `Promise` to let users know that they are talking about a specific promise, but then they use _promise_ so that the text is not obfuscated with too many `Promise`s.
Another nice thing about using both forms is that the second word, _promise_, can be easily 
changed by cases, like in the previous sentence, "promises" would be easier to read, but like in the previous sentence, sometimes we need to use `Promise` to be more specific. 😄

> Every cart service function returns a Promise with a parsed response.

Author: I think the good idea is to apply `Promise` when talking about this specific concept.

You could extend the Cart summary widget to let buyers navigate from this view, 
for example, to checkout, or back to shopping, by adding respective buttons.

### Cart events

When cart data is changed or loaded, the `ibexa-cart:cart-data-changed` event is triggered on `body`.
The reference to the Cart is sent in the event's `detail`.

```js
document.body.addEventListener(
    'ibexa-cart:cart-data-changed',
    ({ detail: { cart } }) => {
        refreshMyWidget(cart);
    },
    false,
);
```

### Cart service 

The Cart package provides `Ibexa\Contracts\Cart\CartServiceInterface` Symfony service, 
which is the entrypoint for calling the [backend API](cart_api.md).

> You can import service using the following code:

Reviewer: ```suggestion
You can import the service using the following code:
```

> You can import services using the following code:

Reviewer 2: ```suggestion
You can import service using the following code:
```

```js
import * as cartService from '@ibexa-cart/src/bundle/Resources/public/js/service/cart';
```

> Use the the service in your code as follows:

Reviewer: ```suggestion
Use the service in your code as follows:
```

```js
cartService.deleteCartEntry(cartIdentifier, entryIdentifier);
```

> Every cart service function returns a `Promise` with a parsed response.

Reviewer: ```suggestion
Every cart service function returns a `Promise` object with a parsed response.
```

When the request is not `OK`, it can throw an error with the response `statusText`.

- `loadUserCarts(ownerId)` - loads 10 user carts
- `loadCartSummary(cartIdentifier)` - load cart summary data
- `createCart(currencyCode)` - creates a new cart
- `deleteCart(cartIdentifier)` - deletes the cart
- `createCartEntry(cartIdentifier, productCode, quantity)` - creates a new cart entry for the specified product
- `updateProductQuantity(cartIdentifier, entryIdentifier, quantity)` - updates product quantity to a new value
- `deleteCartEntry(cartIdentifier, entryIdentifier)` - deletes cart entry
- `emptyCart(cartIdentifier)` - empties cart by removing all entries, returns Promise

To import and initialize cart (without extending it or passing any options), add the following:

```js
import Cart from '@ibexa-cart/src/bundle/Resources/public/js/component/cart';

new Cart();
```

### Change request before sending

> Before the request `ibexa-cart:prepare-request` is sent, an event is triggered on `document`, so you can interchange request objects:

Reviewer 2: ```suggestion
Before every request is sent by `cartService`, the `ibexa-cart:prepare-request` event is triggered on `document`, so you can change request object by assigning a new one to `detail.request`:
```
